### PR TITLE
feat: account suspense

### PR DESCRIPTION
**Pull request by zzmp**
## Description

Shows a connecting state for returning users who have a wallet that should reconnect. This state is based on their last displayed name (address or ENS name), and will persist until the connection is re-established (including ENS, if applicable) or fails.

## Screen capture

https://github.com/Uniswap/interface/assets/5403956/b70d26fe-9cfb-4b20-b909-ecc98965090c

## Test plan
Manually tested with a selectedWallet and without.

### CODE REVIEW

File: src/components/Web3Status/index.tsx [hunk 1]
```diff
@@ -128,7 +129,10 @@ const StyledConnectButton = styled.button`
 function Web3StatusInner() {
   const switchingChain = useAppSelector((state) => state.wallets.switchingChain)
   const ignoreWhileSwitchingChain = useCallback(() => !switchingChain, [switchingChain])
-  const { account, connector } = useLast(useWeb3React(), ignoreWhileSwitchingChain)
+  const connectionReady = useConnectionReady()
+  const activeWeb3 = useWeb3React()
+  const lastWeb3 = useLast(useWeb3React(), ignoreWhileSwitchingChain)
```
Comment (just-toby) [inline src/components/Web3Status/index.tsx:134]: i know you didn't add `useLast` but the name is a little confusing. i think `useMostRecent` would make more sense

File: src/connection/eagerlyConnect.ts [hunk 2]
```diff
@@ -1,8 +1,23 @@
...
+    (onStoreChange) => {
+      if (connectionReady instanceof Promise) {
+        connectionReady.finally(onStoreChange)
+      }
+      return () => undefined
+    },
+    () => connectionReady === true
+  )
+}
+
```
Comment (just-toby) [inline src/connection/eagerlyConnect.ts:20]: this is an interesting pattern. is this only needed because the connected state is updating outside the react lifecycle?

**Review by just-toby — APPROVED** [review]

File: src/connection/eagerlyConnect.ts (hunk 2 above)
Comment (zzmp) [reply src/connection/eagerlyConnect.ts:20]: yes, exactly that. `useSyncExternalStore` let's the react lifecycle subscribe to external updates and trigger renders accordingly.

File: src/components/Web3Status/index.tsx (hunk 1 above)
Comment (zzmp) [reply src/components/Web3Status/index.tsx:134]: I'd support auditing `useLast` and `usePrevious` for both performance and naming. I think they each only have a couple of uses.

File: src/components/Web3Status/index.tsx [hunk 3]
```diff
@@ -173,6 +177,12 @@ function Web3StatusInner() {
         </Web3StatusConnected>
       </TraceEvent>
     )
+  } else if (!connectionReady) {
+    return (
+      <Web3StatusConnected disabled={true}>
+        <LoaderV3 size="24px" />
```
Comment (cbachmeier) [inline src/components/Web3Status/index.tsx:183]: this works well for mobile but we should consider using a pill for screenWidths > 1100px to prevent the trashing of going from the small circular loader to the full wallet pill
thrash from the spinner to full account pill
thrashCircle


Loader pills used on token explore for reference
Screenshot 2023-09-20 at 9 38 21 AM

File: src/connection/eagerlyConnect.ts [hunk 4]
```diff
@@ -26,11 +41,15 @@ const selectedWallet = toConnectionType(localStorage.getItem(selectedWalletKey)
 if (selectedWallet) {
   const selectedConnection = getConnection(selectedWallet)
   if (selectedConnection) {
-    connect(selectedConnection.connector, selectedWallet).then((connected) => {
-      if (!connected) {
-        // only clear the persisted wallet type if it failed to connect.
-        localStorage.removeItem(selectedWalletKey)
-      }
-    })
+    connectionReady = connect(selectedConnection.connector, selectedWallet)
```
Comment (cbachmeier) [inline src/connection/eagerlyConnect.ts:44]: do you need to set `connectionReady = Promise<void>`? Could you instead `let connectionReady = false` and then just update it to `true` on a successful connect? That would simplify the `if` statement in `useConnectionReady`

File: src/components/Web3Status/index.tsx (hunk 3 above)
Comment (zzmp) [reply src/components/Web3Status/index.tsx:183]: A loader pill would not eliminate layout shift / jank / thrashing, because the address is not displayed in a monospaced font, and the width is unknown until connection occurs. we can "cheat", but we can't necessarily get it exactly right.

Comment (zzmp) [reply src/components/Web3Status/index.tsx]: added a "cheat", lmk what you think.

File: src/components/Web3Status/index.tsx [hunk 5]
```diff
@@ -146,6 +146,15 @@ function Web3StatusInner() {
 
   const { hasPendingActivity, pendingActivityCount } = usePendingActivity()
 
+  const displayName = useMemo(() => (account ? ENSName || shortenAddress(account) : null), [account, ENSName])
+  const lastDisplayName = useRef(localStorage.getItem(selectedWalletDisplayKey) ?? null)
+  useEffect(() => {
+    if (displayName) {
+      // It takes a moment to load the ENSName, so set the lastDisplayName (for next time) to the account to reduce layout shift.
+      localStorage.setItem(selectedWalletDisplayKey, shortenAddress(account))
```
Comment (cbachmeier) [inline src/components/Web3Status/index.tsx:154]: why is lastDisplayName set to `shortenAddress(account)` and not `displayName`?

File: src/components/Web3Status/index.tsx [hunk 6]
```diff
@@ -181,7 +190,13 @@ function Web3StatusInner() {
   } else if (!connectionReady) {
     return (
       <Web3StatusConnected disabled={true}>
-        <LoaderV3 size="24px" />
+        <IconWrapper size={24}>
+          <LoaderV3 size="24px" />
+        </IconWrapper>
+        <AddressAndChevronContainer>
+          {/* Render the last known wallet address to prevent a layout shift on initial connection. */}
+          <Text style={{ visibility: 'hidden' }}>{lastDisplayName.current}</Text>
```
Comment (cbachmeier) [inline src/components/Web3Status/index.tsx:198]: Definitely hacky but effective and looks a lot cleaner loading

File: src/connection/eagerlyConnect.ts [hunk 7]
```diff
@@ -43,9 +45,14 @@ if (selectedWallet) {
...
-          // only clear the persisted wallet type if it failed to connect.
-          localStorage.removeItem(selectedWalletKey)
+        if (!connected) throw new FailedToConnect()
+      })
+      .catch((error) => {
+        // Clear the persisted wallet type if it failed to connect.
+        localStorage.removeItem(selectedWalletKey)
+        // Log it if it threw an unknown error.
+        if (!(error instanceof FailedToConnect)) {
+          console.error(error)
```
Comment (cbachmeier) [inline src/connection/eagerlyConnect.ts:55]: do we do anything with console logged errors? Should we instead log it to Sentry?

**Review by cbachmeier — APPROVED** [review]

File: src/components/Web3Status/index.tsx (hunk 5 above)
Comment (zzmp) [reply src/components/Web3Status/index.tsx:154]: Because the account address will load before ens. We could tighten this up more but this was a first attempt. 

See inline comment above this line.

File: src/connection/eagerlyConnect.ts (hunk 7 above)
Comment (zzmp) [reply src/connection/eagerlyConnect.ts:55]: A user may just be denying a connection. I think this is the same behavior as before.

**Review by just-toby — DISMISSED** [review]

File: src/components/Web3Status/index.tsx [hunk 8]
```diff
@@ -147,13 +148,36 @@ function Web3StatusInner() {
...
+        type: connection.type,
+        display: displayName,
+        displayType: ENSName ? ConnectionDisplayType.ENSName : ConnectionDisplayType.Address,
+      }
+      setConnectionMeta(meta)
     }
-  }, [account, displayName])
+  }, [ENSName, account, connection.type, displayName])
+  const isDisplayNameLoading = Boolean(
+    account && meta.current?.displayType === ConnectionDisplayType.ENSName && ENSLoading
```
Comment (cbachmeier) [inline src/components/Web3Status/index.tsx]: Is there a situation where we would be loading an ENSName without an `account` or `meta.current?.displayType !== ConnectionDisplayType.ENSName`?

File: cypress/utils/user-state.ts [hunk 9]
```diff
@@ -13,7 +14,12 @@ export const DISCONNECTED_WALLET_USER_STATE: Partial<UserState> = { selectedWall
 export function setInitialUserState(win: Cypress.AUTWindow, state: UserState) {
   // Selected wallet should also be reflected in localStorage, so that eager connections work.
   if (state.selectedWallet) {
-    win.localStorage.setItem('selected_wallet', state.selectedWallet)
+    win.localStorage.setItem(
```
Comment (cbachmeier) [inline cypress/utils/user-state.ts:17]: could we use the `setConnectionMeta` fn here?

Comment (zzmp) [reply cypress/utils/user-state.ts:17]: we can't without passing the window object, which complicates the setter. I chose to just export the key, rather than make the setter more complex.

File: src/components/Web3Status/index.tsx (hunk 8 above)
Comment (zzmp) [reply src/components/Web3Status/index.tsx]: not that i can think of, no

**Review by cbachmeier — APPROVED** [review]

File: src/components/Web3Status/index.tsx [hunk 10]
```diff
@@ -141,6 +153,37 @@ function Web3StatusInner() {
 
   const { hasPendingActivity, pendingActivityCount } = usePendingActivity()
 
+  const displayName = useMemo(() => (account ? ENSName || shortenAddress(account) : null), [account, ENSName])
```
Comment (just-toby) [inline src/components/Web3Status/index.tsx:156]: i think we usually use `??` instead of `||`

File: src/components/Web3Status/index.tsx [hunk 11]
```diff
@@ -141,6 +153,37 @@ function Web3StatusInner() {
...
+  const meta = useRef(getConnectionMeta())
```
Comment (just-toby) [inline src/components/Web3Status/index.tsx:157]: using a ref for this makes the flow hard to follow. maybe if you wrapped it in a `useCurrentMeta` hook which handled updating when the connection loads, this file would be easier to follow

File: src/components/Web3Status/index.tsx (hunk 10 above)
Comment (just-toby) [reply src/components/Web3Status/index.tsx:156]: also we usually use `undefined` instead of `null` now

File: src/components/Web3Status/index.tsx [hunk 12]
```diff
@@ -141,6 +153,37 @@ function Web3StatusInner() {
...
+      const meta: ConnectionMeta = {
+        type: connection.type,
+        address: account,
+        ENSName: ENSName ?? undefined,
+      }
+      setConnectionMeta(meta)
+    }
+  }, [ENSName, account, connection.type, displayName])
+
+  const isENSNameLoading = Boolean(meta.current?.address === account && meta.current?.ENSName && ENSLoading)
```
Comment (just-toby) [inline src/components/Web3Status/index.tsx:169]: this variable name is also confusing to me, because it includes more than just the loading state. 

is `meta` just supposed to be the cached wallet information? if you can clarify that with a couple comments or name changes then i think this file will be alot easier to follow

**Review by just-toby — CHANGES_REQUESTED** [review]

**Review by just-toby — APPROVED** [review]

**Review by cbachmeier — APPROVED** [review]
